Add timit recipe (including the code scripts and the docs) for icefall

luomingshuang opened the pull request:
In this timit recipe, it nearly includes all required files (the code scripts and the docs). I place two dirs (`tdnn_lstm_ctc` and `tdnn_ligru_ctc`) in this recipe.  In the tdnn_lstm_ctc, the PER is `19.71%`. and in the tdnn_ligru_ctc, the PER is `17.66%`.

In this version, I have deleted the splits_dir and including it in prepare_timit.py in lhotse.

About the colab notebook for the pretrained.py in the docs, I will add it until it is merged.

csukuangfj commented on egs/timit/ASR/local/compute_fbank_timit.py:
```diff
@@ -0,0 +1,97 @@
+                recordings=m["recordings"],
+                supervisions=m["supervisions"],
+            )
+            if "TRAIN" in partition:
```
```suggestion
            if partition == "TRAIN":
```

csukuangfj commented on egs/timit/ASR/local/prepare_lexicon.py:
```diff
@@ -0,0 +1,102 @@
+        for load_dict in load_dicts:
+            text = load_dict["text"]
+            # list the phone units and filter the empty item
+            phones_list = list(filter(None, text.split()))
```
```suggestion
            phones_list = text.split()
```

luomingshuang replied:
Em...this maybe keep it. Some texts are like follow：
`sil ay f  g sil`--->`['sil', 'ay', 'f', '', 'g', 'sil']`, so we have to filter this guy `''` .

csukuangfj commented on egs/timit/ASR/local/prepare_lexicon.py:
```diff
@@ -0,0 +1,102 @@
+    Return:
+      The lexicon.txt file and the train.text in lang_dir.
+    """
+    phones = set([])
```
```suggestion
    phones = set()
```

csukuangfj commented on egs/timit/ASR/prepare.sh:
```diff
@@ -0,0 +1,154 @@
+      --read-symbol-table="data/lang_phone/words.txt" \
+      --disambig-symbol='#0' \
+      --max-order=4 \
+      $dl_dir/lm/lm_4_gram_tgmed.arpa > data/lm/G_4_gram.fst.txt
```
What is the meaning of `tgmed` in a 4-gram arpa. I thought `tg` means tri-gram.

csukuangfj commented on egs/timit/ASR/tdnn_ligru_ctc/decode.py:
```diff
@@ -0,0 +1,493 @@
+    if params.method in ["nbest-rescoring", "whole-lattice-rescoring"]:
+        if not (params.lm_dir / "G_4_gram.pt").is_file():
+            logging.info("Loading G_4_gram.fst.txt")
+            logging.warning("It may take 8 minutes.")
```
Does it also take 8 minutes for TIMIT? I think the LM in Librispeech is much larger than the one in TIMIT.

csukuangfj reviewed (commented):
👍

luomingshuang commented:
Oh.&nbsp;A&nbsp;careless..


---Original---
From: &quot;Fangjun ***@***.***&gt;
Date: Tue, Nov 9, 2021 22:58 PM
To: ***@***.***&gt;;
Cc: &quot;Mingshuang ***@***.******@***.***&gt;;
Subject: Re: [k2-fsa/icefall] Add timit recipe (including the code scripts andthe docs) for icefall (PR #114)




 
@csukuangfj commented on this pull request.
 


 
In egs/timit/ASR/local/compute_fbank_timit.py:
 &gt; +    ) +    assert manifests is not None + +    extractor = Fbank(FbankConfig(num_mel_bins=num_mel_bins)) + +    with get_executor() as ex:  # Initialize the executor only once. +        for partition, m in manifests.items(): +            if (output_dir / f&quot;cuts_{partition}.json.gz&quot;).is_file(): +                logging.info(f&quot;{partition} already exists - skipping.&quot;) +                continue +            logging.info(f&quot;Processing {partition}&quot;) +            cut_set = CutSet.from_manifests( +                recordings=m[&quot;recordings&quot;], +                supervisions=m[&quot;supervisions&quot;], +            ) +            if &quot;TRAIN&quot; in partition:  ⬇️ Suggested change -            if "TRAIN" in partition: +            if partition == "TRAIN":  
 
In egs/timit/ASR/local/prepare_lexicon.py:
 &gt; + +    Return: +      The lexicon.txt file and the train.text in lang_dir. +    &quot;&quot;&quot; +    phones = set([]) + +    supervisions_train = Path(manifests_dir) / &quot;supervisions_TRAIN.json&quot; +    lexicon = Path(lang_dir) / &quot;lexicon.txt&quot; + +    logging.info(f&quot;Loading {supervisions_train}!&quot;) +    with open(supervisions_train, &quot;r&quot;) as load_f: +        load_dicts = json.load(load_f) +        for load_dict in load_dicts: +            text = load_dict[&quot;text&quot;] +            # list the phone units and filter the empty item +            phones_list = list(filter(None, text.split()))  ⬇️ Suggested change -            phones_list = list(filter(None, text.split())) +            phones_list = text.split()  
 
In egs/timit/ASR/local/prepare_lexicon.py:
 &gt; + +    return parser.parse_args() + + +def prepare_lexicon(manifests_dir: str, lang_dir: str): +    &quot;&quot;&quot; +    Args: +      manifests_dir: +        The manifests directory, e.g., data/manifests. +      lang_dir: +        The language directory, e.g., data/lang_phone. + +    Return: +      The lexicon.txt file and the train.text in lang_dir. +    &quot;&quot;&quot; +    phones = set([])  ⬇️ Suggested change -    phones = set([]) +    phones = set()  
 
In egs/timit/ASR/prepare.sh:
 &gt; +  if [ ! -f data/lm/G_3_gram.fst.txt ]; then +    # It is used in building HLG +    python3 -m kaldilm \ +      --read-symbol-table=&quot;data/lang_phone/words.txt&quot; \ +      --disambig-symbol='#0' \ +      --max-order=3 \ +      $dl_dir/lm/lm_3_gram_tgmed.arpa &gt; data/lm/G_3_gram.fst.txt +  fi + +  if [ ! -f data/lm/G_4_gram.fst.txt ]; then +    # It is used for LM rescoring +    python3 -m kaldilm \ +      --read-symbol-table=&quot;data/lang_phone/words.txt&quot; \ +      --disambig-symbol='#0' \ +      --max-order=4 \ +      $dl_dir/lm/lm_4_gram_tgmed.arpa &gt; data/lm/G_4_gram.fst.txt  
What is the meaning of tgmed in a 4-gram arpa. I thought tg means tri-gram.
 
 
In egs/timit/ASR/tdnn_ligru_ctc/decode.py:
 &gt; + +    logging.info(f&quot;device: {device}&quot;) + +    HLG = k2.Fsa.from_dict( +        torch.load(f&quot;{params.lang_dir}/HLG.pt&quot;, map_location=&quot;cpu&quot;) +    ) +    HLG = HLG.to(device) +    assert HLG.requires_grad is False + +    if not hasattr(HLG, &quot;lm_scores&quot;): +        HLG.lm_scores = HLG.scores.clone() + +    if params.method in [&quot;nbest-rescoring&quot;, &quot;whole-lattice-rescoring&quot;]: +        if not (params.lm_dir / &quot;G_4_gram.pt&quot;).is_file(): +            logging.info(&quot;Loading G_4_gram.fst.txt&quot;) +            logging.warning(&quot;It may take 8 minutes.&quot;)  
Does it also take 8 minutes for TIMIT? I think the LM in Librispeech is much larger than the one in TIMIT.
 
&mdash;
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

csukuangfj commented:
Is it ready to merge?

luomingshuang commented:
> Is it ready to merge?

Oh, let me add a README.md before that.

luomingshuang commented:
Now, I think it is ready to merge.

csukuangfj commented on docs/source/recipes/index.rst:
```diff
@@ -10,8 +10,10 @@ We may add recipes for other tasks as well in the future.
 .. Other recipes are listed in a alphabetical order.
```
Could you also change `README.md` at the root directory to add the results for `TIMIT` recipe.

luomingshuang replied:
Done!  And will finish the colab running after merging.

csukuangfj commented:
Thanks!
